### PR TITLE
[Bugfix] Fix request disconnect check when BaseHTTPMiddleware is present

### LINKED ISSUE #10087: [Bug]: Requests aren't aborted when client disconnects if user provided a middleware
In https://github.com/vllm-project/vllm/pull/7111/, @njhill did a great job to make sure requests are aborted if they are cancelled by the client. This works great and can really ease the load on the model engine in cases of timeouts / other client disconnects.

Unfortunately, this mechanism doesn't work if a `BaseHTTPMiddleware` is added to the server, which is the case when a use provides a function to add as middleware (via the `--middleware` cmd arg of `vllm serve`). In this case, some weird starlette behavior causes `request.is_disconnected()` to return `False`, even if the client disconnects. This is discussed in https://github.com/encode/starlette/discussions/2094.
One of the suggested solutions in that discussion (for example, in https://github.com/encode/starlette/discussions/2094#discussioncomment-9084737 and https://github.com/encode/starlette/discussions/2094#discussioncomment-7576379) is to change the mechanism used to check if the request was actually disconnected. Copied verbatim from that discussion, use this helper instead of `request.is_disconnected()`:

> ``` python
> async def my_is_disconnected(request: Request) -> bool:
>     assert hasattr(request, '_is_disconnected'), "private API in starlette changed"
>     assert isinstance(request._is_disconnected, bool), "private API in starlette changed"
> 
>     if request._is_disconnected:
>         return True
> 
>     message: Message = {}
> 
>     # this timeout may need to be tweaked.
>     # Ideally request.receive() is non-blocking, in which case the timeout doesn't matter.
>     # But the ASGI spec seems to imply it should be a blocking callable, as it doesn't specify
>     # what should be returned in case of no new messages. In this situation you can return an
>     # empty dict, but given that it's against the spec it seems inadvisable.
>     with anyio.move_on_after(0.01):
>         message = await request.receive()
> 
>     if message.get("type") == "http.disconnect":
>         request._is_disconnected = True
> 
>     return request._is_disconnected
> 
> ```

In vLLM, the `is_cancelled` callable is passed to `iterate_with_cancellation` or to `merge_async_iterators`. In all cases, `request.is_disconnected` is used as that callable. So it should be pretty easy to replace it with the helper function instead. The only issue I see is with the timeout in `anyio.move_on_after`.

So WDYT? Is it possible to tweak the `is_cancelled` callable so requests will be aborted upon client disconnect also when the user provides a middleware function?

### PR DESCRIPTION
**Pull request by tomeras91**
This PR fixes https://github.com/vllm-project/vllm/issues/10087, by using a different method to check if a request has disconnected (instead of starlette's `Request.is_disconnected()`, which doesn't work as expected in case a BaseHTTPMiddleware is added to the server - see here - https://github.com/encode/starlette/discussions/2094).

The PR uses the method suggested in https://github.com/encode/starlette/discussions/2094#discussioncomment-9084737

Added a test that fails on `main`, and is fixed with this PR

### Benchmark
Model: Qwen/Qwen2.5-1.5B-Instruct
Hardware: single H100
Serve run command: `vllm serve Qwen/Qwen2.5-1.5B-Instruct`

Endpoint: `v1/completions`
(benchmark run command: `python3 benchmark_serving.py --model Qwen/Qwen2.5-1.5B-Instruct --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json`)

| branch | throughput (requests/s) | mean TTFT (msec) | mean TPOT (msec) | Total generated tokens |
|--------|--------|--------|--------|--------|
| main (commit 82c73fd5104e010c2c98820f3e761e1e4f36c135) | 52.12 | 6117.88 | 24.92 | 189868 |
| pr (commit  0ba0ff8791b62a7104cc7387a863a03e40c1d493) | 52.52 | 6064.14 | 24.20 | 190241 |

Conclusion: The fix doesn't hurt performance at all. I guess this is because `request.recieve()` used [here](https://github.com/vllm-project/vllm/pull/11096/files#diff-dab7693bd00a09e22d39aee684a7e419aa358a47c4bd20df33d44f5adf60d304R402) is generally non-blocking

### CODE REVIEW

**Comment by tomeras91** [comment]
CC @njhill

File: vllm/utils.py [hunk 1]
```diff
@@ -370,6 +373,40 @@ def _next_task(iterator: AsyncGenerator[T, None],
...
+    # but given that it's against the spec it seems inadvisable.
+    with anyio.move_on_after(0.01):
+        message = await request.receive()
+
+    if message.get("type") == "http.disconnect":
+        request._is_disconnected = True
+
+    return request._is_disconnected
+
+
```
Comment (DarkLight1337) [inline vllm/utils.py:409]: IMO it's more appropriate to put this inside the base serving class since it's only really being used there.

Comment (tomeras91) [reply vllm/utils.py:409]: It's used also in the legacy `vllm/entrypoints/api_server.py`. Also, this is completely stateless so I'm not really sure it should be a part of `OpenAIServing`. I don't mind moving it elsewhere just not sure the base serving class is the correct place.
Maybe create a new utils file under `vllm/entrypints`?

Comment (DarkLight1337) [reply vllm/utils.py:409]: Oh, I see. Maybe we can place it inside `vllm/entrypoints/utils.py`.

Comment (tomeras91) [reply vllm/utils.py]: Done

**Comment by DarkLight1337** [comment]
As per offline discussion, we have decided to fix it via #11190 instead.

**Pull request closed without merging**